### DISCUSSION TITLE
nextjs with zustand-v4 (zustand/context migration guide)

### DISCUSSION THREAD

**Discussion by benx-wang** [pmndrs/zustand#1180] [General] [answered]
I am trying zustand with next.js based on [with-zustand-example](https://github.com/vercel/next.js/tree/canary/examples/with-zustand) that depends on `creatContext` function.

As [`creatContext` will likely be deprecated in v4 and removed in v5](https://docs.pmnd.rs/zustand/zustand-v3-create-context) , 
So what is the right pattern to update the example ?

**Comment by dai-shi** [comment, marked as answer]
Good point.
We have an example in https://github.com/pmndrs/zustand/releases/tag/v4.0.0, but not migration guide.

## zustand/context migration guide

Previously, we did something like this:

```js
import create from 'zustand'
import createContext from 'zustand/context'

const MyContext = createContext()

const createMyStore = () => create(...)

const App = () => (
  <MyContext.Provider createStore={createMyStore}>
    ...
  </MyContext.Provider>
)

const Component = () => {
  const slice = MyContext.useStore(selector)
  ...
}
```

With v4, it can be the following:

```js
import { createStore, useStore } from 'zustand'
import { createContext, useContext } from 'react'

const MyContext = createContext()

const createMyStore = () => createStore(...)

const App = () => {
  const store = useRef(createMyStore()).current // or any better solution
  return (
    <MyContext.Provider value={store}>
      ...
    </MyContext.Provider>
  )
}

const Component = () => {
  const store = useContext(MyContext)
  const slice = useStore(store, selector)
  ...
}
```

**Reply by matthewdavi** [reply]
Is there an example of how to create a typed `useBoundStore` hook that encapsulates the
 ```
         const store = useContext(MyContext); 

         const slice = useStore(store, selector) 
```

pattern? I imagine this could get annoying to type over and over again.

**Reply by dai-shi** [reply]
For typing context, can you try something like this?
```ts
const createMyStore = () => createStore(...)

const MyContext = createContext<ReturnType<typeof createMyStore> | null>(null)
```

**Reply by mrtnpro** [reply]
This worked perfectly for me, thanks @dai-shi

**Reply by matthewdavi** [reply]
Hello, can you provide typing for the useContextStore at the bottom. I'm not sure how to type the selector to get the same behavior as a regular non-context store. The example below does not produce proper type inference. 

```ts


import { createContext, useState, useContext } from 'react';
import { createStore, useStore } from 'zustand';
interface ContextStore {
  count: number;
  inc: () => void;
}
const createMyStore = () =>
  createStore<ContextStore>((set, get) => ({
    count: 0,
    inc: () => set((state) => ({ ...state, count: state.count + 1 })),
  }));

const MyContext = createContext<ReturnType<typeof createMyStore> | null>(null);


function Wrapper({ children }: { children: React.ReactNode }) {
  const [store] = useState(() => createMyStore());
  return <MyContext.Provider value={store}>{children}</MyContext.Provider>;
}

function useContextStore(
  selector: (state: ContextStore) => typeof state[keyof typeof state]
) {
  const store = useContext(MyContext)!;
  const value = useStore(store, selector);
  return value;
}
```

**Reply by matthewdavi** [reply]
I think it might be this but I'm not sure. Can you take a look? Thanks. [Stackblitz link](https://stackblitz.com/edit/vitejs-vite-ezfnft?file=src%2FuseContextStore.tsx&terminal=dev)
```ts

function useContextSelector<T extends ContextStore[keyof ContextStore]>(
  selector: (state: ContextStore) => T
) {
  const store = useContextStore();
  const value = useStore(store, selector);
  return value;
}
```

**Reply by dai-shi** [reply]
Here's what I would do:

```tsx
import React, { createContext, useRef, useContext } from "react";
import { createStore, useStore } from "zustand";

interface ContextStore {
  count: number;
  inc: () => void;
}

const createMyStore = () =>
  createStore<ContextStore>((set) => ({
    count: 0,
    inc: () => set((state) => ({ count: state.count + 1 })),
  }));

const MyContext = createContext<ReturnType<typeof createMyStore> | null>(null);

export function Wrapper({ children }: { children: React.ReactNode }) {
  const storeRef = useRef<ReturnType<typeof createMyStore>>();
  if (!storeRef.current) {
    storeRef.current = createMyStore();
  }
  return (
    <MyContext.Provider value={storeRef.current}>{children}</MyContext.Provider>
  );
}

export function useContextStore<T>(selector: (state: ContextStore) => T) {
  const store = useContext(MyContext);
  if (store === null) {
    throw new Error("Missing Wrapper in the tree");
  }
  const value = useStore(store, selector);
  return value;
}
```

TS playground: https://tsplay.dev/mpL9Mm

**Reply by cuppachino** [reply]
I really love this @dai-shi. I think it would be helpful if this was included in the docs with the optional equalityFn argument included.
```diff
export function useContextStore<T>(
  selector: (state: ContextStore) => T,
+  equalityFn?: (a: T, b: T) => boolean
) {
  const store = useContext(MyContext);
  if (store === null) {
    throw new Error('Missing Wrapper in the tree');
  }
- const value = useStore(store, selector);
+ const value = useStore(store, selector, equalityFn);
  return value;
}
```

**Reply by dai-shi** [reply]
Would you like to open a  PR to add `./docs/zustand-v4-context-usage.md` doc?

**Reply by cuppachino** [reply]
> Would you like to open a PR to add `./docs/zustand-v4-context-usage.md` doc?

definitely! I'll try to get it done before tomorrow.

**Reply by razgraf** [reply]
Has this been included in the docs yet? It looks like the context example in [readme](https://github.com/pmndrs/zustand#react-context) is lacking some nuances (e.g. `useRef`) and the TS examples don't include it either.

**Reply by cuppachino** [reply]
Yes @razgraf https://github.com/pmndrs/zustand/blob/main/docs/guides/initialize-state-with-props.md

**Reply by maxgurewitz** [reply]
Hi all. Are there any plans to update the [next.js example](https://github.com/vercel/next.js/blob/deprecated-main/examples/with-zustand/lib/store.js#L3) to use the new api? @Munawwar . 

Thank you.

**Reply by prabak** [reply]
Hi,

I am trying to use nextjs with the latest version of zustand. Do we need to use special handling for server side vs client side on nextjs if the store is bound to a page. That is, there is no global store but store created for specific pages via:

```js
function BearProvider({ children, ...props }: BearProviderProps) {
  const storeRef = useRef<BearStore>()
  if (!storeRef.current) {
    storeRef.current = useCreateBearStore(props);
  }
  return (
    <BearContext.Provider value={storeRef.current}>
      {children}
    </BearContext.Provider>
  )
}
```
thank you.

**Reply by dai-shi** [reply]
That looks fine. We just recommend `createStore` and `useStore` over `create` in the context usage.

**Reply by Munawwar** [reply]
> Hi all. Are there any plans to update the [next.js example](https://github.com/vercel/next.js/blob/deprecated-main/examples/with-zustand/lib/store.js#L3) to use the new api? @Munawwar .
> 
> Thank you.

I haven't used next in a while now (and haven't upgraded zustand too in a while). I guess its up to you all to keep the community moving.
("The power is yours!" says captain planet and exits)

**Comment by karltaylor** [comment]
How would this work when copying the ssr example? https://github.com/vercel/next.js/blob/canary/examples/with-zustand/lib/store.js

**Reply by arvinxx** [reply]
you can still use createContext method with [zustand-utils](https://github.com/arvinxx/zustand-utils)

**Reply by prabak** [reply]
Someone created a PR: https://github.com/vercel/next.js/pull/46911

**Comment by mhfaust** [comment]
Has anyone done this with TS? The most basic example following the docs doesn't compile. 
Problems:
- createContext requires an initial value (not provided in sample code, but that one's easy enough to fix)
- The context returned from createContext isn't accepted by useContext -- "Argument of type 'TestShape' is not assignable to parameter of type `WithReact<StoreApi<unknown>>`"

```
import { createContext, useContext, ReactNode, useRef } from "react";
import { createStore, useStore } from "zustand";

type TestShape = {
  key: string;
}

const createTestStore = () => createStore<TestShape>((set) => {
  return {
    key: '',
    setKey: (key: string) => set({ key })
  }
});

const TestContext = createContext<TestShape>({
  key: ''
};

export const Provider = ({ children }: { children: ReactNode}) => {
  const gameStore = useRef(createTestStore()).current;
  return (
    <TestContext.Provider value={gameStore.getState()}>
      {children}
    </TestContext.Provider>
    
  );
};

const useTestStore = (selector: any) => useStore(useContext(TestContext), selector);
```

**Reply by cuppachino** [reply]
you need to curry `createStore<TestShape>()((set) => ....`

**EDIT** Here's a full example

```ts
import { createContext, useContext, ReactNode, useRef } from "react";
import { createStore, useStore } from "zustand";

type TestShape = {
  key: string;
}
type TestShapeActions = {
  setKey: (key: string) => void;
}

const createTestStore = () => createStore<TestShape & TestShapeActions>()((set) => {
  return {
    key: '',
    setKey: (key: string) => set((state) => ({ ...state, key }))
  }
});

type TestStore = ReturnType<typeof createTestStore>;

const TestContext = createContext<TestStore | null>(null);

export const Provider = ({ children }: { children: ReactNode }) => {
  const gameStoreRef = useRef<TestStore>();
  if (!gameStoreRef.current) {
    gameStoreRef.current = createTestStore();
  }
  return (
    <TestContext.Provider value={gameStoreRef.current}>
      {children}
    </TestContext.Provider>

  );
};

export const Consumer = () => {
  const storeApi = useContext(TestContext);
  if (!storeApi) throw new Error("Missing TestContext.Provider in the tree")
  const setKey = useStore(storeApi, (state) => state.setKey);
  const key = useStore(storeApi, (state) => state.key);
  return (
    <div>
      <button type="button" onClick={() => setKey("myKey")}>Set Key</button>
      <div>Key: {key}</div>
    </div>)
}

function App() {
  return (
    <div className="App">
      <Provider>
        <Consumer />
      </Provider>
    </div>
  )
}

export default App
```

Is anything [from the guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/initialize-state-with-props.md) broken?

**Comment by riccardolardi** [comment]
The example https://github.com/vercel/next.js/tree/canary/examples/with-zustand seems to work fine for me, thanks a lot for making this. One question though: if I wanted to re-initialize / update the store on route change (after hydration) based on new data from `getStaticProps`, how would I do that?

In `myPage.tsx`
```jsx
export const getStaticProps: GetStaticProps = async (context) => {
  const { mainMenuData } = myFetchRequest()...
  const store = initializeStore({ mainMenuData })

  return {
    props: {
      initialStoreState: JSON.parse(JSON.stringify(store.getState())),
    },
  }
}
```

This will work fine for first request but will not update `mainMenuData` after route change since:

In `store-provider.tsx`
```jsx
const StoreProvider = ({ children, ...props }: PropsWithChildren) => {
  const storeRef = useRef<StoreType>()

  if (!storeRef.current) { // this will block updating the values
    storeRef.current = initializeStore(props)
  }

  return <Provider value={storeRef.current}>{children}</Provider>
}
```

Can I just remove that check `if (!storeRef.current)` and set `initializeStore(props)` every time the props change?

**Comment by Igorlialko** [comment]
I created npm package for Provider zustand
https://www.npmjs.com/package/next-zustand
Happy coding bro !)
